qe/drivers.py

```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+    # Set initial guess vectors and subspace dimension
+    dim_S = min(n_eig * n_guess, n)
+    V_guess = np.eye(n, dim_S)
+    V_k = np.c_[V_k, V_guess]
+
+    n_newvecs = dim_S  # No. of vectors added is initial subspace dimension
+    converged = False
+    restart = 0
+    for k in range(1, max_iter):
+        # Get new trial vectors and compute new columns of W_k
+        V_new = np.array(V_k[:, -n_newvecs:], dtype="float")
+        W_new = np.dot(H, V_new)
+        W_k = np.c_[W_k, W_new]
+        # Rayleigh-Ritz; Update projected Hamiltonian
+        if (k == 1) or (
+            restart == 1
+        ):  # If first iterate (or following a restart), need to compute full S_k explicitly
+            S_k = np.dot(V_k.T, W_k)
+            restart = 0
+        else:  # Else, append new rows & columns
+            S_new_c = np.dot(V_k[:, :-n_newvecs].T, W_new)
+            S_k = np.c_[S_k, S_new_c]
+            S_new_r = np.c_[np.dot(V_new.T, W_k[:, :-n_newvecs]), np.dot(V_new.T, W_new)]
+            S_k = np.r_[S_k, S_new_r]
+        # Diagonalize S_k and keep n_eig smallest estimates
+        L_k, Y_k = np.linalg.eigh(S_k)
+        L_k = L_k[:n_eig]
+        Y_k = Y_k[:, :n_eig]
+
+        n_newvecs = 0  # Initialize counter; no. of new vectors added to trial subspace
+        X_k = np.dot(V_k, Y_k)  # Pre-compute Ritz vectors (V_k updated each iteration)
+        converged = True  # Reset
+        for j in range(n_eig):
+            # Grab jth eigenpair, each rank computes local portion of jth residual vector
+            l_j, y_j = L_k[j], Y_k[:, j]
+            r_j = np.dot(W_k, y_j) - l_j * X_k[:, j]
+            res = np.linalg.norm(r_j)
+            if res > eps:  # If ||r_j|| > eps, jth eigenpair hasn't converged
+                converged = False
+                # Precondition next trial vector
+                M_j = np.diag(
+                    np.clip(np.reciprocal(D - l_j), a_min=-1e5, a_max=1e5)
+                )  # Build diagonal preconditioner
+                t_j = np.dot(M_j, r_j)
+                # Orthogonalize new trial vector against previous basis vectors using Modified Gram-Schmidt (MG)S
+                t_j = t_j / np.linalg.norm(t_j)
+                for i in range(V_k.shape[1]):  # Iterate through k basis vectors
+                    c_i = np.copy(
+                        np.inner(V_k[:, i], t_j)
+                    )  # Each process computes partial inner-product
+                    t_j -= c_i * V_k[:, i]  # Remove component of t_j in V_k
+
+                norm_tj = np.linalg.norm(t_j)
+                if norm_tj > 1e-4:  # If new vector is `small`, ignore. Avoids ill-conditioning
+                    V_k = np.c_[V_k, t_j / norm_tj]  # Append new vector to trial subspace
+                    n_newvecs += 1
+
+        if converged == True:
             break
-        theta_old = theta[:eig]
+
+        dim_S += n_newvecs  # Update dimension of trial subspace
```

Comment:
Look like we don't need to do the addition if `n_newvecs == 0` , should we do this in the `if q<= dims banch` ?

Reply:
It does get used in the check for that branch.  I could do

'''
if q - n_newvecs <= dim_S:  # Collapose trial basis
                self.print_master("q <= dim_S: {dim_S}, restarting Davidson's")
                dim_S, n_newvecs, V_ik, W_ik = self.parallel_restart(
                    dim_S, n_eig, n_newvecs, X_ik, V_ik, W_ik
                )
                restart = True 
                dim_S += n_newvecs
'''

Reply:
Actually, I think this need to be outside, for the cases where we don't enter the restart branches